addons/pos_viva_wallet/models/pos_payment_method.py

```diff
@@ -0,0 +1,229 @@
+
+        retry += 1
+        if resp.text and json.loads(resp.text).get('detail') == 'Could not validate credentials' and retry <= 3:
+            self._bearer_token()
+            self._retrieve_session_id(data_webhook, session)
+        elif resp.status_code == 200:
+            data = json.loads(resp.text)
+            if data.get('success'):
+                data.update({'pos_session_id': pos_session_id})
+                data.update({'data_webhook': data_webhook})
+                self.viva_wallet_latest_response = data
+                self._send_notification(data)
+        elif retry > 4:
+            self._send_notification({'error': 'Could not validate credentials'})
+        else:
+            self._send_notification(
+                {'error': _(
+                    "There are some issues between us and Viva Wallet, try again later. %s",
+                    json.loads(resp.text).get('detail')
+                    )}
+                )
+
+    def _send_notification(self, data):
+        # Send a notification to the point of sale channel to indicate that the transaction are finish
+        pos_session_sudo = request.env["pos.session"].sudo().browse(int(data.get('pos_session_id', False)))
+        if pos_session_sudo:
+            self.env['bus.bus']._sendone(pos_session_sudo._get_bus_channel_name(), 'VIVA_WALLET_LATEST_RESPONSE', pos_session_sudo.config_id.id)
+
+    def viva_wallet_send_payment_request(self, data, session=None, retry=0):
+        if not self.env.user.has_group('point_of_sale.group_pos_user'):
+            raise AccessError(_("Do not have access to fetch token from Viva Wallet"))
+
+        if not session:
+            session = self._get_session()
+        session.headers.update({'Authorization': f"Bearer {self.viva_wallet_bearer_token}"})
+        try:
+            resp = session.post(f"{self._viva_wallet_api_get_endpoint()}/ecr/v1/transactions:sale", json=data, timeout=TIMEOUT)
+        except requests.exceptions.RequestException:
+            return {'error': _("There are some issues between us and Viva Wallet, try again later. (viva_wallet_send_payment_request)")}
+
+        retry += 1
+        if resp.text and json.loads(resp.text).get('detail') == 'Could not validate credentials' and retry <= 3:
+            self._bearer_token()
+            return self.viva_wallet_send_payment_request(data, session, retry)
+        elif resp.status_code == 200:
+            return {'success': resp.status_code}
+        elif retry > 4:
+            return {'error': 'Could not validate credentials'}
+        else:
+            return {'error': _("There are some issues between us and Viva Wallet, try again later. %s", json.loads(resp.text).get('detail'))}
+
+    def viva_wallet_send_payment_cancel(self, data, session=None, retry=0):
+        if not self.env.user.has_group('point_of_sale.group_pos_user'):
+            raise AccessError(_("Do not have access to fetch token from Viva Wallet"))
```

Comment:
Try to hint how to resolve an issue in your error messages. For instance, instead of "you are not allowed", you can say "access is restricted to X" or "only Y users are allowed to"